### DISCUSSION TITLE
How to create longhorn volume via kubernetes manifest

### DISCUSSION THREAD

**Discussion by viceice** [longhorn/longhorn#2190] [Q&A] [answered]
I'm trying to auto create a longhorn volume via a kubernetes persistant volume manifest (terraform) but it will always in pending state.

```hcl
resource "kubernetes_persistent_volume_claim" "maria_data" {
  metadata {
    name      = kubernetes_persistent_volume.maria.metadata.0.name
    namespace = var.namespace
  }
  spec {
    access_modes       = ["ReadWriteOnce"]
    storage_class_name = "longhorn-static"
    volume_name        = kubernetes_persistent_volume.maria.metadata.0.name

    resources {
      requests = {
        storage = "2.5Gi"
      }
    }
  }
}


resource "kubernetes_persistent_volume" "maria-data" {
  metadata {
    name = "maria-data"
  }

  spec {
    access_modes       = ["ReadWriteOnce"]
    storage_class_name = "longhorn-static"
    volume_mode        = "Filesystem"

    capacity = {
      storage = "2.5Gi"
    }

    persistent_volume_source {
      csi {
        driver        = "driver.longhorn.io"
        fs_type       = "ext4"
        volume_handle = "maria-data"

        volume_attributes = {
          numberOfReplicas    = "3"
          staleReplicaTimeout = "2880"
        }
      }
    }
  }
}
```

use [this](https://github.com/longhorn/longhorn/blob/a6b624b889ee7f4c2a233fadc972c513563244a6/examples/csi/example_pv.yaml#L1) as base

**Comment by shuo-wu** [comment, marked as answer]
Kubernetes PVC and PV are different from Longhorn volume. A Longhorn volume is the underlying of the PV here.  Hence, you need to manually create Longhorn volume `volume_handle = "maria-data"` via Longhorn UI before applying these 2 manifests.

If you don't want to use Longhorn UI. You can ask Kubernetes to help you create PV and Longhorn volumes based on a PVC manifest: https://github.com/longhorn/longhorn/blob/master/examples/simple_pvc.yaml

**Reply by viceice** [reply]
I know the simply pvc, but if the pv and longhorn volume name are not kown, it will be a lot harder to restore, as i need to move / copy the data from old to new pv.

Or is there a simply way to recreate the pvc from longhorn backup without defining the pv manually? If i define the pv manually i loose the auto volume expand feature. :-/

**Reply by shuo-wu** [reply]
The previous PVC and PV names will be still stored in the volume after you restored the Longhorn volume from a backup. Then you can use Longhorn UI to create PVC and PV with the old names. 

Actually, I am confused about your use case. As long as you know which Longhorn volume you want to operate, you can directly get the PVC and PV name on the volume detail page. If there is no PVC and PV for a Longhorn volume, you can use Longhorn UI to create them, or write the manifests like you show above to create them.

**Reply by viceice** [reply]
OK, i'll try that in a test cluster.

- create and backup a volume with data
- recreate kubernetes cluster
- restore volume 
- check data

🙃

**Comment by red-avtovo** [comment]
Simple pvc indeed creates a longhorn volume, but unfortunately, you have no control over its name. I tried to provide the pv name or its handler, but eventually, I got stuck

**Reply by shuo-wu** [reply]
You need to either rely on PVC to create a PV and Longhorn volume (with random names), or create everything by yourself.
